eclair-core/src/main/scala/fr/acinq/eclair/crypto/TransportHandler.scala

```diff
@@ -118,162 +119,191 @@ class TransportHandler[T: ClassTag](keyPair: KeyPair, rs: Option[ByteVector], co
-            context.parent ! HandshakeCompleted(connection, self, remoteNodeId)
-            val nextStateData = WaitingForListenerData(Encryptor(ExtendedCipherState(enc, ck)), Decryptor(ExtendedCipherState(dec, ck), ciphertextLength = None, remainder))
-            goto(WaitingForListener) using nextStateData
-
-          case (writer, _, None) => {
-            writer.write(ByteVector.empty) match {
-              case (reader1, message, None) => {
-                // we're still in the middle of the handshake process and the other end must first received our next
-                // message before they can reply
-                require(remainder.isEmpty, "unexpected additional data received during handshake")
-                connection ! Tcp.Write(buf(TransportHandler.prefix +: message))
-                stay using HandshakeData(reader1, remainder)
-              }
-              case (_, message, Some((enc, dec, ck))) => {
-                connection ! Tcp.Write(buf(TransportHandler.prefix +: message))
-                val remoteNodeId = PublicKey(writer.rs)
-                remoteNodeId_opt = Some(remoteNodeId)
-                context.parent ! HandshakeCompleted(connection, self, remoteNodeId)
-                val nextStateData = WaitingForListenerData(Encryptor(ExtendedCipherState(enc, ck)), Decryptor(ExtendedCipherState(dec, ck), ciphertextLength = None, remainder))
-                goto(WaitingForListener) using nextStateData
+    handleExceptions {
+      case Event(Tcp.Received(data), HandshakeData(reader, buffer)) =>
+        connection ! Tcp.ResumeReading
+        log.debug("received {}", ByteVector(data))
+        val buffer1 = buffer ++ data
+        if (buffer1.length < expectedLength(reader))
+          stay using HandshakeData(reader, buffer1)
+        else {
+          if (buffer1.head != TransportHandler.prefix) throw InvalidTransportPrefix(ByteVector(buffer1))
+
+          val (payload, remainder) = buffer1.tail.splitAt(expectedLength(reader) - 1)
+
+          reader.read(ByteVector.view(payload.asByteBuffer)) match {
+            case (writer, _, Some((dec, enc, ck))) =>
+              val remoteNodeId = PublicKey(writer.rs)
+              remoteNodeId_opt = Some(remoteNodeId)
+              context.parent ! HandshakeCompleted(connection, self, remoteNodeId)
+              val nextStateData = WaitingForListenerData(Encryptor(ExtendedCipherState(enc, ck)), Decryptor(ExtendedCipherState(dec, ck), ciphertextLength = None, remainder))
+              goto(WaitingForListener) using nextStateData
+
+            case (writer, _, None) => {
+              writer.write(ByteVector.empty) match {
+                case (reader1, message, None) => {
+                  // we're still in the middle of the handshake process and the other end must first received our next
+                  // message before they can reply
+                  require(remainder.isEmpty, "unexpected additional data received during handshake")
```

Comment:
Another require to remove! Death to them all!

Reply:
Good catch!